File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    answer = 0
    
    # 최소, 최대값
    left = min(times)
    right = n * max(times)
    
    # 이분 탐색
    while left <= right:
        people = 0      # 심사한 사람 수
        mid = (left+right) // 2
        for time in times:
            people += mid // time
            if people >= n:
                break
    
        if people >= n:
            right = mid-1
            answer = mid
        else:
            left = mid+1
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long left = 1;
        long right = (long) times[0]*n;
        
        while (left < right) {
            long mid = (left + right) / 2;
            if (enough(mid, times, n)) right = mid;
            else left = mid + 1;
        }
        
        return left;
    }
    
    boolean enough(long t, int[] times, int n) {
        long done = 0;
        for (int time : times) {
            done += t/time;
            if (done >= n) return true;
        }
        return false;
    }
}